### PR TITLE
MaterializedMySQL flush relevant tables only (RDS support)

### PR DESCRIPTION
**Pull request by poweroftrue**
### Changelog category (leave one):
- Improvement

### Changelog entry (a user-readable short description of the changes that goes to CHANGELOG.md):
Add MaterializedMySQL support for RDS and other managed databases that don't support SUPER privilege.

This will solve issues like https://github.com/ClickHouse/ClickHouse/issues/27302 https://github.com/ClickHouse/ClickHouse/issues/23541.

I think it's worth mentioning in docs that attaching Clickhouse (flushing tables) to the master will block insert/modify/DDL operations on the tables causing downtime, doing it on a replica will stop replication even if one table was passed to materialized_mysql_tables_list. That's because the SQL thread will wait for the lock effectively pausing replication to all tables.

MaterializedMySQL is awesome huge thanks to everybody working on it ❤️  

We're running CH to replicate from a multi-terabyte database and it's crushing it! The master running on `db.r5.16xlarge` (64 vCPU, 512 GB ram), and Clickhouse is running on `c5.2xlarge` (8 vCPU, 16 GB ram) and it's keeping up with sub-second replication lag 💪  (we're replicating from a replica not from the master directly).

We ran `sum`s and `count()`s on CH and the data is much more consistent and accurate than the ETL to Redshift that we were using.

This is inspired by https://github.com/debezium/debezium/pull/198 https://github.com/debezium/debezium/pull/177 https://github.com/debezium/debezium/pull/190

### CODE REVIEW

**Comment by tavplubix** [comment]
In general LGTM, but integration tests failures are probably related

**Comment by poweroftrue** [comment]
I'll check them 👍

**Comment by poweroftrue** [comment]
@tavplubix I checked the tests, mainly the reason for the failures is that we sometimes try to `SETTINGS materialized_mysql_tables_list = ' a,c,d'` while in MySQL the only tables are `a,b` (to test if we get the correct filtered tables based on the `materialized_mysql_tables_list`).

That worked when we were locking all tables anyway, now with this change, we only lock the required tables, which causes MySQL to fail to acquire the lock due to c,d tables are not in MySQL: `Error Code: 1146. Table 'd' doesn't exist` when we run `FLUSH TABLES a,c,d WITH READ LOCK;`.

So, we have two ways to address this, either fix the tests to only include existing tables or change the CH code to ignore tables that don't exist.

I think we should fix the tests because if a user added a table list and some tables are not there it shouldn't fail silently. It will lead to a lot of debugging to know the typo'd table name and why it wasn't replicating. 

@tavplubix let me know what you think and I'll implement it 👍

**Comment by tavplubix** [comment]
I think it was supposed to ignore non-existing table. Because `MaterializedMySQL` replicates DDL queries as well and some tables may be created or dropped in MySQL later.

**Comment by adhasahar97** [comment]
Hi. Any update on this?

**Comment by poweroftrue** [comment]
@mergify update

**Comment by poweroftrue** [comment]
Hello @adhasahar97 I'll give it another shot in the next couple of days.

In the meantime, you can use my fork or fork a new one with the same lines, and compile it.
My compiled fork is working fine for about two months now.

I'll take another look at the failing integration tests although testing it locally (running the same commands as in the integration test works just fine).

File: src/Databases/MySQL/MaterializeMetadata.cpp [hunk 1]
```diff
@@ -293,9 +294,39 @@ void MaterializeMetadata::startReplication(
...
+            /// is not supported even for the master user. However, we can use FLUSH TABLES table_names WITH READ LOCK.
+            std::unordered_set<String> base_tables_list;
+
+            /// Get only base tables (not views).
+            mysqlxx::UseQueryResult result_list = connection->query("SHOW FULL TABLES IN " + database + " WHERE Table_type = 'BASE TABLE'").use();
+
+            while (mysqlxx::Row row = result_list.fetch())
+                base_tables_list.insert(row[0].getString());
+
+            connection->query("FLUSH TABLES `" + boost::algorithm::join(base_tables_list, "`,`") + "` WITH READ LOCK;").execute();
```
Comment (tavplubix) [inline src/Databases/MySQL/MaterializeMetadata.cpp:310]: It's incorrect, quoting like this never works (what will happen if a table name contains backtick?). Use `backQuoteIfNeed`.

**Comment by tavplubix** [comment]
This PR is abandoned, closing

**Pull request closed without merging**